File: 백준/Gold/14502. 연구소/연구소.py
import copy
from itertools import combinations
from collections import deque

dx = [0,0,-1,1]
dy = [1,-1,0,0]

def bfs(tmp_board):
    q = deque()
    visited = [[0 for _ in range(M)] for _ in range(N)]
    for i in virus_lst:
        q.append(i)
    while q:
        x,y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < M:
                if visited[nx][ny] == 0 and tmp_board[nx][ny] == 0:
                    q.append((nx,ny))
                    visited[nx][ny] = 1
                    tmp_board[nx][ny] = 2

    res = 0
    for i in range(N):
        for j in range(M):
            if tmp_board[i][j] == 0:
                res+=1

    return res


N,M = map(int,input().split())
# 0: 빈칸, 1: 벽, 2: 바이러스
board = [list(map(int,input().split())) for _ in range(N)]

#바이러스 위치
virus_lst = []
for i in range(N):
    for j in range(M):
        if board[i][j] == 2:
            virus_lst.append((i,j))

wall = []
for i in range(N):
    for j in range(M):
        if board[i][j] != 1 and board[i][j] != 2:
            wall.append((i,j))
#벽 좌표 후보
wall_candi = []
for i in combinations(wall,3):
    wall_candi.append(i)

ans = 0
for i in wall_candi:
    tmp_board = copy.deepcopy(board)
    #벽 세우기
    for j in i:
        tmp_board[j[0]][j[1]] = 1

    res = bfs(tmp_board)
    ans = max(ans,res)

print(ans)

File: 백준/Gold/14502. 연구소/연구소.java
import java.util.*;
import java.io.*;

public class Main {
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};

    static int bfs(int[][] tmp_board, List<Pair<Integer, Integer>> virus_lst, int N, int M) {
        Queue<Pair<Integer, Integer>> q = new LinkedList<>();
        int[][] visited = new int[N][M];
        for (Pair<Integer, Integer> i : virus_lst) {
            q.add(i);
        }
        while (!q.isEmpty()) {
            Pair<Integer, Integer> pair = q.poll();
            int x = pair.getKey();
            int y = pair.getValue();
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (0 <= nx && nx < N && 0 <= ny && ny < M) {
                    if (visited[nx][ny] == 0 && tmp_board[nx][ny] == 0) {
                        q.add(new Pair<>(nx, ny));
                        visited[nx][ny] = 1;
                        tmp_board[nx][ny] = 2;
                    }
                }
            }
        }

        int res = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (tmp_board[i][j] == 0) {
                    res++;
                }
            }
        }
        return res;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[][] board = new int[N][M];

        // 바이러스 위치
        List<Pair<Integer, Integer>> virus_lst = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
                if (board[i][j] == 2) {
                    virus_lst.add(new Pair<>(i, j));
                }
            }
        }

        List<Pair<Integer, Integer>> wall = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] != 1 && board[i][j] != 2) {
                    wall.add(new Pair<>(i, j));
                }
            }
        }

        List<List<Pair<Integer, Integer>>> wall_candi = new ArrayList<>();
        for (int i = 0; i < wall.size(); i++) {
            for (int j = i + 1; j < wall.size(); j++) {
                for (int k = j + 1; k < wall.size(); k++) {
                    List<Pair<Integer, Integer>> combination = new ArrayList<>();
                    combination.add(wall.get(i));
                    combination.add(wall.get(j));
                    combination.add(wall.get(k));
                    wall_candi.add(combination);
                }
            }
        }

        int ans = 0;
        for (List<Pair<Integer, Integer>> combination : wall_candi) {
            int[][] tmp_board = new int[N][M];
            for (int i = 0; i < N; i++) {
                tmp_board[i] = Arrays.copyOf(board[i], M);
            }
            for (Pair<Integer, Integer> j : combination) {
                tmp_board[j.getKey()][j.getValue()] = 1;
            }
            int res = bfs(tmp_board, virus_lst, N, M);
            ans = Math.max(ans, res);
        }

        System.out.println(ans);
    }
    
    static class Pair<X, Y> {
        private final X first;
        private final Y second;

        public Pair(X first, Y second) {
            this.first = first;
            this.second = second;
        }

        public X getKey() {
            return first;
        }

        public Y getValue() {
            return second;
        }
    }
}